File: c++/프로그래머스/Level2/위클리챌린지/9주차_전력망둘로나누기.py
def init(parents) :
    for i in range(0, len(parents)) :
        parents[i] = i
    return

def Find(parents, v) :
    if parents[v] == v :
        return v
    parents[v] = Find(parents, parents[v])
    return parents[v]

def Union(parents, v1, v2) :
    p1 = Find(parents, v1)
    p2 = Find(parents, v2)

    if p1 == p2 :
        return
    elif p1 < p2 :
        parents[p2] = p1
    elif p1 > p2 :
        parents[p1] = p2
    return

def Start(wires, parents, visited) :
    for i in range(0, len(wires)) :
        if visited[i]==1 :
            continue
        v1 = wires[i][0]-1
        v2 = wires[i][1]-1
        
        if Find(parents, v1) != Find(parents, v2) :
            Union(parents, v1, v2)
    
    for i in range(0, len(parents)) :
        Find(parents, i)
    
    cnt = 0 
    for i in range(0, len(parents)) :
        if parents[i] == 0 :
            cnt += 1
    
    return abs(cnt-(len(parents)-cnt))
    
def solution(n, wires):
    answer = 102
    parents = [0 for _ in range(n)]
    visited = [0 for _ in range(n)]
    
    for i in range(0, len(wires)) :
        visited[i] = 1
        init(parents)
        answer = min(answer, Start(wires, parents, visited))
        visited[i] = 0
    
    return answer

File: Source Code/c++/프로그래머스/Level2/위클리챌린지/9주차_전력망둘로나누기.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int Find(int parents[], int v){
    if(parents[v]==v){
        return v;
    }
    parents[v] = Find(parents, parents[v]);
    return parents[v];
}

void Union(int parents[], int v1, int v2){
    int p1 = Find(parents, v1);
    int p2 = Find(parents, v2);
    
    if(p1==p2){
        return;
    }
    else if(p1<p2){
        parents[p2] = p1;
    }
    else if(p1>p2){
        parents[p1] = p2;
    }
    return;
}

int solution(int n, vector<vector<int>> wires) {
    int answer = 99999;
    int parents[102];
    int sz = wires.size();
    
    for(int i=0; i<sz; i++){
        for(int j=0; j<=n; j++){
            parents[j] = j;
        }
        
        for(int j=0; j<sz; j++){
            if(i==j)
                continue;
            if(Find(parents, wires[j][0])!=Find(parents, wires[j][1])){
                Union(parents, wires[j][0], wires[j][1]);
            }
        }
        
        for(int j=1; j<=n; j++){
            Find(parents, j);
        }
        
        int cnt = 0;
        for(int j=1; j<=n; j++){
            if(parents[j]==1){
                cnt++;
            }
        }
        answer = min(answer, abs(cnt-(n-cnt)));
    }
    return answer;
}